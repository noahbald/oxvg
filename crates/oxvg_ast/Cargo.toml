[package]
name = "oxvg_ast"
version = "0.0.1-beta.1"
edition = "2021"
license.workspace = true
repository.workspace = true
authors.workspace = true
description = "Trait representation for concrete ASTs"

[lints]
workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[features]
default = ["markup5ever", "roxmltree", "visitor"]
markup5ever = ["dep:rcdom", "dep:xml5ever"]
roxmltree = ["dep:roxmltree"]
serialize = []
selectors = ["dep:selectors", "dep:cssparser", "dep:precomputed-hash"]
visitor = ["dep:bitflags"]
style = ["dep:lightningcss", "dep:cssparser_lightningcss", "dep:smallvec"]

[dependencies]
anyhow = { workspace = true, optional = true }
bitflags = { workspace = true, optional = true }
cfg-if = { workspace = true }
cssparser = { workspace = true, optional = true }
cssparser_lightningcss = { package = "cssparser", version = "0.33", optional = true }
derive_more = { workspace = true, features = ["debug"] }
itertools = { workspace = true }
lightningcss = { workspace = true, optional = true }
log = { workspace = true }
markup5ever = { workspace = true }
precomputed-hash = { workspace = true, optional = true }
rcdom = { workspace = true, optional = true }
roxmltree = { workspace = true, optional = true }
# ensure track_alloc_size for wasm compat
servo_arc = { version = "0.4", features = ["track_alloc_size"] }
selectors = { workspace = true, optional = true }
smallvec = { version = "1.13.2", optional = true }
string_cache = { workspace = true }
tendril = { workspace = true }
typed-arena = { workspace = true }
xml5ever = { workspace = true, optional = true }
