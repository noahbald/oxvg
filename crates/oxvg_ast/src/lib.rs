#[macro_use]
extern crate bitflags;

#[cfg(feature = "style")]
#[macro_use]
extern crate smallvec;

pub mod atom;
pub mod attribute;
pub mod class_list;
pub mod document;
pub mod element;
pub mod implementations;
pub mod name;
pub mod node;

#[cfg(feature = "visitor")]
pub mod visitor;

#[cfg(feature = "parse")]
pub mod parse;

#[cfg(feature = "serialize")]
pub mod serialize;

#[cfg(feature = "selectors")]
pub mod selectors;

#[cfg(feature = "style")]
pub mod style;
