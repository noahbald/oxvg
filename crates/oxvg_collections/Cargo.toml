[package]
name = "oxvg_collections"
version = "0.0.1-alpha.1"
edition = "2021"
license.workspace = true
repository.workspace = true
authors.workspace = true
description = "selectors implementation for rcdom"

[lints]
workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
lazy_static = { workspace = true }
log = { workspace = true }
phf = { workspace = true }
regex = { workspace = true }

[dev-dependencies]
oxvg_ast = { workspace = true }
