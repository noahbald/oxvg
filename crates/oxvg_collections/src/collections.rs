use std::collections::{BTreeMap, BTreeSet};

pub trait Group<'a> {
    fn matches<T>(&self, value: T) -> bool
    where
        T: Into<&'a str>,
    {
        self.set().contains(value.into())
    }

    fn set(&self) -> &'a BTreeSet<&'static str>;
}

#[derive(Hash, PartialEq, Eq)]
pub enum ElementGroup {
    Animation,
    Descriptive,
    Shape,
    Structural,
    PaintServer,
    NonRendering,
    Container,
    TextContent,
    TextContentChild,
    LightSource,
    FilterPrimitive,
}

impl<'a> Group<'a> for ElementGroup {
    fn set(&self) -> &'a BTreeSet<&'static str> {
        match self {
            Self::Animation => &ANIMATION,
            Self::Descriptive => &DESCRIPTIVE,
            Self::Shape => &SHAPE,
            Self::Structural => &STRUCTURAL,
            Self::PaintServer => &PAINT_SERVER,
            Self::NonRendering => &NON_RENDERING,
            Self::Container => &CONTAINER,
            Self::TextContent => &TEXT_CONTENT,
            Self::TextContentChild => &TEXT_CONTENT_CHILD,
            Self::LightSource => &LIGHT_SOURCE,
            Self::FilterPrimitive => &FILTER_PRIMITIVE,
        }
    }
}

// Element groups
lazy_static! {
    static ref ANIMATION: BTreeSet<&'static str> = BTreeSet::from([
        "animate",
        "animateColor",
        "animateMotion",
        "animateTransform",
        "set"
    ]);
    static ref DESCRIPTIVE: BTreeSet<&'static str> = BTreeSet::from(["desc", "metadata", "title"]);
    static ref SHAPE: BTreeSet<&'static str> =
        BTreeSet::from(["circle", "ellipse", "line", "path", "polygon", "polyline", "rect"]);
    static ref STRUCTURAL: BTreeSet<&'static str> =
        BTreeSet::from(["defs", "g", "svg", "symbol", "use"]);
    static ref PAINT_SERVER: BTreeSet<&'static str> = BTreeSet::from([
        "hatch",
        "linearGradient",
        "meshGradient",
        "pattern",
        "radialGradient",
        "solidColor"
    ]);
    static ref NON_RENDERING: BTreeSet<&'static str> = BTreeSet::from([
        "clipPath",
        "filter",
        "linearGradient",
        "marker",
        "mask",
        "pattern",
        "radialGradient",
        "solidColor",
        "symbol"
    ]);
    static ref CONTAINER: BTreeSet<&'static str> = BTreeSet::from([
        "a",
        "defs",
        "foreignObject",
        "g",
        "marker",
        "mask",
        "missing-glyph",
        "pattern",
        "svg",
        "switch",
        "symbol"
    ]);
    static ref TEXT_CONTENT: BTreeSet<&'static str> = BTreeSet::from([
        "a",
        "altGlyph",
        "altGlyphDef",
        "alyGlyphItem",
        "glyph",
        "glyphRef",
        "text",
        "textPath",
        "tref",
        "tspan"
    ]);
    static ref TEXT_CONTENT_CHILD: BTreeSet<&'static str> =
        BTreeSet::from(["alyGlyph", "textPath", "tref", "tspan"]);
    static ref LIGHT_SOURCE: BTreeSet<&'static str> = BTreeSet::from([
        "feDiffuseLighting",
        "feDistantLight",
        "fePointLight",
        "feSpecularLighting",
        "feSpotLight"
    ]);
    static ref FILTER_PRIMITIVE: BTreeSet<&'static str> = BTreeSet::from([
        "feBlend",
        "feColorMatrix",
        "feComponentTransfer",
        "feComposite",
        "feConvolveMatrix",
        "feDiffuseLighting",
        "feDisplacementMap",
        "feDropShadow",
        "feFlood",
        "feFuncA",
        "feFuncB",
        "feFuncG",
        "feFuncR",
        "feGaussianBlur",
        "feImage",
        "feMerge",
        "feMergeNode",
        "feMorphology",
        "feOffset",
        "feSpecularLighting",
        "feTile",
        "feTurbulence"
    ]);
}

#[derive(Hash, PartialEq, Eq)]
pub enum AttrsGroups {
    AnimationAddition,
    AnimationAttributeTarget,
    AnimationEvent,
    AnimationTiming,
    AnimationValue,
    ConditionalProcessing,
    Core,
    GraphicalEvent,
    Presentation,
    XLink,
    DocumentEvent,
    DocumentElementEvent,
    GlobalEvent,
    FilterPrimitive,
    TransferFunction,
}

impl<'a> Group<'a> for AttrsGroups {
    fn set(&self) -> &'a BTreeSet<&'static str> {
        match self {
            Self::AnimationAddition => &ANIMATION_ADDITION,
            Self::AnimationAttributeTarget => &ANIMATION_ATTRIBUTE_TARGET,
            Self::AnimationEvent => &ANIMATION_EVENT,
            Self::AnimationTiming => &ANIMATION_TIMING,
            Self::AnimationValue => &ANIMATION_VALUE,
            Self::ConditionalProcessing => &CONDITIONAL_PROCESSING,
            Self::Core => &CORE,
            Self::GraphicalEvent => &GRAPHICAL_EVENT,
            Self::Presentation => &PRESENTATION,
            Self::XLink => &X_LINK,
            Self::DocumentEvent => &DOCUMENT_EVENT,
            Self::DocumentElementEvent => &DOCUMENT_EVEMENT_EVENT,
            Self::GlobalEvent => &GLOBAL_EVENT,
            Self::FilterPrimitive => &FILTER_PRIMITIVE,
            Self::TransferFunction => &TRANSFER_FUNCTION,
        }
    }
}

// Attrs groups
lazy_static! {
    pub static ref ANIMATION_ADDITION: BTreeSet<&'static str> =
        BTreeSet::from(["additive", "accumulate"]);
    pub static ref ANIMATION_ATTRIBUTE_TARGET: BTreeSet<&'static str> =
        BTreeSet::from(["attributeType", "attributeName"]);
    pub static ref ANIMATION_EVENT: BTreeSet<&'static str> =
        BTreeSet::from(["onbegin", "onend", "onrepeat", "onload"]);
    pub static ref ANIMATION_TIMING: BTreeSet<&'static str> = BTreeSet::from([
        "begin",
        "dur",
        "end",
        "fill",
        "max",
        "min",
        "repeatCount",
        "repeatDur",
        "restart",
    ]);
    pub static ref ANIMATION_VALUE: BTreeSet<&'static str> = BTreeSet::from([
        "by",
        "calcMode",
        "from",
        "keySplines",
        "keyTimes",
        "to",
        "values",
    ]);
    pub static ref CONDITIONAL_PROCESSING: BTreeSet<&'static str> =
        BTreeSet::from(["requiredExtensions", "requiredFeatures", "systemLanguage",]);
    pub static ref CORE: BTreeSet<&'static str> =
        BTreeSet::from(["id", "tabindex", "xml:base", "xml:lang", "xml:space"]);
    pub static ref GRAPHICAL_EVENT: BTreeSet<&'static str> = BTreeSet::from([
        "onactivate",
        "onclick",
        "onfocusin",
        "onfocusout",
        "onload",
        "onmousedown",
        "onmousemove",
        "onmouseout",
        "onmouseover",
        "onmouseup",
    ]);
    pub static ref PRESENTATION: BTreeSet<&'static str> = BTreeSet::from([
        "alignment-baseline",
        "baseline-shift",
        "clip-path",
        "clip-rule",
        "clip",
        "color-interpolation-filters",
        "color-interpolation",
        "color-profile",
        "color-rendering",
        "color",
        "cursor",
        "direction",
        "display",
        "dominant-baseline",
        "enable-background",
        "fill-opacity",
        "fill-rule",
        "fill",
        "filter",
        "flood-color",
        "flood-opacity",
        "font-family",
        "font-size-adjust",
        "font-size",
        "font-stretch",
        "font-style",
        "font-variant",
        "font-weight",
        "glyph-orientation-horizontal",
        "glyph-orientation-vertical",
        "image-rendering",
        "letter-spacing",
        "lighting-color",
        "marker-end",
        "marker-mid",
        "marker-start",
        "mask",
        "opacity",
        "overflow",
        "paint-order",
        "pointer-events",
        "shape-rendering",
        "stop-color",
        "stop-opacity",
        "stroke-dasharray",
        "stroke-dashoffset",
        "stroke-linecap",
        "stroke-linejoin",
        "stroke-miterlimit",
        "stroke-opacity",
        "stroke-width",
        "stroke",
        "text-anchor",
        "text-decoration",
        "text-overflow",
        "text-rendering",
        "transform-origin",
        "transform",
        "unicode-bidi",
        "vector-effect",
        "visibility",
        "word-spacing",
        "writing-mode",
    ]);
    pub static ref X_LINK: BTreeSet<&'static str> = BTreeSet::from([
        "xlink:actuate",
        "xlink:arcrole",
        "xlink:href",
        "xlink:role",
        "xlink:show",
        "xlink:title",
        "xlink:type",
    ]);
    pub static ref DOCUMENT_EVENT: BTreeSet<&'static str> =
        BTreeSet::from(["onabort", "onerror", "onresize", "onscroll", "onunload", "onzoom",]);
    pub static ref DOCUMENT_EVEMENT_EVENT: BTreeSet<&'static str> =
        BTreeSet::from(["oncopy", "oncut", "onpaste"]);
    pub static ref GLOBAL_EVENT: BTreeSet<&'static str> = BTreeSet::from([
        "oncancel",
        "oncanplay",
        "oncanplaythrough",
        "onchange",
        "onclick",
        "onclose",
        "oncuechange",
        "ondblclick",
        "ondrag",
        "ondragend",
        "ondragenter",
        "ondragleave",
        "ondragover",
        "ondragstart",
        "ondrop",
        "ondurationchange",
        "onemptied",
        "onended",
        "onerror",
        "onfocus",
        "oninput",
        "oninvalid",
        "onkeydown",
        "onkeypress",
        "onkeyup",
        "onload",
        "onloadeddata",
        "onloadedmetadata",
        "onloadstart",
        "onmousedown",
        "onmouseenter",
        "onmouseleave",
        "onmousemove",
        "onmouseout",
        "onmouseover",
        "onmouseup",
        "onmousewheel",
        "onpause",
        "onplay",
        "onplaying",
        "onprogress",
        "onratechange",
        "onreset",
        "onresize",
        "onscroll",
        "onseeked",
        "onseeking",
        "onselect",
        "onshow",
        "onstalled",
        "onsubmit",
        "onsuspend",
        "ontimeupdate",
        "ontoggle",
        "onvolumechange",
        "onwaiting",
    ]);
    pub static ref FILTER_PRIMITIVE_ATTRS: BTreeSet<&'static str> =
        BTreeSet::from(["x", "y", "width", "height", "result"]);
    pub static ref TRANSFER_FUNCTION: BTreeSet<&'static str> = BTreeSet::from([
        "amplitude",
        "exponent",
        "intercept",
        "offset",
        "slope",
        "tableValues",
        "type",
    ]);
}

impl AttrsGroups {
    pub fn defaults(&self) -> Option<&BTreeMap<&'static str, &'static str>> {
        match self {
            Self::Core => Some(&CORE_DEFAULTS),
            Self::Presentation => Some(&PRESENTATION_DEFAULTS),
            Self::TransferFunction => Some(&TRANSFER_FUNCTION_DEFAULTS),
            _ => None,
        }
    }
}

// Attrs groups defaults
lazy_static! {
    static ref CORE_DEFAULTS: BTreeMap<&'static str, &'static str> =
        BTreeMap::from([("xml:space", "default")]);
    static ref PRESENTATION_DEFAULTS: BTreeMap<&'static str, &'static str> = BTreeMap::from([
        ("clip", "auto"),
        ("clip-path'", "none"),
        ("clip-rule'", "nonzero"),
        ("mask", "none"),
        ("opacity", "1"),
        ("stop-color'", "#000"),
        ("stop-opacity'", "1"),
        ("fill-opacity'", "1"),
        ("fill-rule'", "nonzero"),
        ("fill", "#000"),
        ("stroke", "none"),
        ("stroke-width'", "1"),
        ("stroke-linecap'", "butt"),
        ("stroke-linejoin'", "miter"),
        ("stroke-miterlimit'", "4"),
        ("stroke-dasharray'", "none"),
        ("stroke-dashoffset'", "0"),
        ("stroke-opacity'", "1"),
        ("paint-order'", "normal"),
        ("vector-effect'", "none"),
        ("display", "inline"),
        ("visibility", "visible"),
        ("marker-start'", "none"),
        ("marker-mid'", "none"),
        ("marker-end'", "none"),
        ("color-interpolation'", "sRGB"),
        ("color-interpolation-filters'", "linearRGB"),
        ("color-rendering'", "auto"),
        ("shape-rendering'", "auto"),
        ("text-rendering'", "auto"),
        ("image-rendering'", "auto"),
        ("font-style'", "normal"),
        ("font-variant'", "normal"),
        ("font-weight'", "normal"),
        ("font-stretch'", "normal"),
        ("font-size'", "medium"),
        ("font-size-adjust'", "none"),
        ("kerning", "auto"),
        ("letter-spacing'", "normal"),
        ("word-spacing'", "normal"),
        ("text-decoration'", "none"),
        ("text-anchor'", "start"),
        ("text-overflow'", "clip"),
        ("writing-mode'", "lr-tb"),
        ("glyph-orientation-vertical'", "auto"),
        ("glyph-orientation-horizontal'", "0deg"),
        ("direction", "ltr"),
        ("unicode-bidi'", "normal"),
        ("dominant-baseline'", "auto"),
        ("alignment-baseline'", "baseline"),
        ("baseline-shift'", "baseline"),
    ]);
    static ref TRANSFER_FUNCTION_DEFAULTS: BTreeMap<&'static str, &'static str> = BTreeMap::from([
        ("slope", "1"),
        ("intercept", "0"),
        ("amplitude", "1"),
        ("exponent", "1"),
        ("offset", "0"),
    ]);
}

lazy_static! {
    pub static ref PATH_ELEMS: BTreeSet<&'static str> =
        BTreeSet::from(["glyph", "missing-glyph", "path"]);
    pub static ref INHERITABLE_ATTRS: BTreeSet<&'static str> = BTreeSet::from([
        "clip-rule",
        "color-interpolation-filters",
        "color-interpolation",
        "color-profile",
        "color-rendering",
        "color",
        "cursor",
        "direction",
        "dominant-baseline",
        "fill-opacity",
        "fill-rule",
        "fill",
        "font-family",
        "font-size-adjust",
        "font-size",
        "font-stretch",
        "font-style",
        "font-variant",
        "font-weight",
        "font",
        "glyph-orientation-horizontal",
        "glyph-orientation-vertical",
        "image-rendering",
        "letter-spacing",
        "marker-end",
        "marker-mid",
        "marker-start",
        "marker",
        "paint-order",
        "pointer-events",
        "shape-rendering",
        "stroke-dasharray",
        "stroke-dashoffset",
        "stroke-linecap",
        "stroke-linejoin",
        "stroke-miterlimit",
        "stroke-opacity",
        "stroke-width",
        "stroke",
        "text-anchor",
        "text-rendering",
        "transform",
        "visibility",
        "word-spacing",
        "writing-mode",
    ]);
    pub static ref PRESENTATION_NON_INHERITABLE_GROUP_ATTRS: BTreeSet<&'static str> =
        BTreeSet::from([
            "clip-path",
            "display",
            "filter",
            "mask",
            "opacity",
            "text-decoration",
            "transform",
            "unicode-bidi",
        ]);
    pub static ref REFERENCES_PROPS: BTreeSet<&'static str> = BTreeSet::from([
        "clip-path",
        "color-profile",
        "fill",
        "filter",
        "marker-end",
        "marker-mid",
        "marker-start",
        "mask",
        "stroke",
        "style",
    ]);
    pub static ref PSEUDO_DISPLAY_STATE: BTreeSet<&'static str> =
        BTreeSet::from(["fullscreen", "modal", "picture-in-picture"]);
    pub static ref PSEUDO_INPUT: BTreeSet<&'static str> = BTreeSet::from([
        "autofill",
        "blank",
        "checked",
        "default",
        "disabled",
        "enabled",
        "in-range",
        "indetermined",
        "invalid",
        "optional",
        "out-of-range",
        "placeholder-shown",
        "read-only",
        "read-write",
        "required",
        "user-invalid",
        "valid",
    ]);
    pub static ref PSEUDO_LINGUISTIC: BTreeSet<&'static str> = BTreeSet::from(["dir", "lang"]);
    pub static ref PSEUDO_LOCATION: BTreeSet<&'static str> = BTreeSet::from([
        "any-link",
        "link",
        "local-link",
        "scope",
        "target-within",
        "target",
        "visited",
    ]);
    pub static ref PSEUDO_RESOURCE_STATE: BTreeSet<&'static str> =
        BTreeSet::from(["playing", "paused"]);
    pub static ref PSEUDO_TIME_DIMENSIONAL: BTreeSet<&'static str> =
        BTreeSet::from(["current", "past", "future"]);
    pub static ref PSEUDO_TREE_STRUCTURAL: BTreeSet<&'static str> = BTreeSet::from([
        "empty",
        "first-child",
        "first-of-type",
        "last-child",
        "last-of-type",
        "nth-child",
        "nth-last-child",
        "nth-last-of-type",
        "nth-of-type",
        "only-child",
        "only-of-type",
        "root",
    ]);
    pub static ref PSEUDO_USER_ACTION: BTreeSet<&'static str> =
        BTreeSet::from(["active", "focus-visible", "focus-within", "focus", "hover",]);
    pub static ref PSEUDO_FUNCTIONAL: BTreeSet<&'static str> =
        BTreeSet::from(["is", "not", "where", "has"]);
}
