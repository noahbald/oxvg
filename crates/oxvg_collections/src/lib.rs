#[macro_use]
extern crate lazy_static;

pub mod allowed_content;
pub mod collections;
pub mod regex;
