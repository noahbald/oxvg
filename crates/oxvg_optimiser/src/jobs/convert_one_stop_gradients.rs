use std::{
    cell::{Cell, RefCell},
    collections::HashSet,
    marker::PhantomData,
};

use derive_where::derive_where;
use itertools::Itertools as _;
use lightningcss::{
    printer::PrinterOptions,
    properties::{
        custom::{CustomProperty, CustomPropertyName, TokenOrValue},
        Property, PropertyId,
    },
    traits::ToCss as _,
};
use oxvg_ast::{
    attribute::Attr,
    element::Element,
    name::Name as _,
    style::{ComputedStyles, Id, PresentationAttr, PresentationAttrId, Static, Style},
    visitor::{Context, ContextFlags, Info, PrepareOutcome, Visitor},
};
use oxvg_collections::collections::{AttrsGroups, COLORS_PROPS};
use serde::{Deserialize, Serialize};

#[cfg(feature = "wasm")]
use tsify::Tsify;

#[cfg_attr(feature = "wasm", derive(Tsify))]
#[cfg_attr(feature = "napi", napi(object))]
#[derive(Deserialize, Serialize, Debug, Default, Clone)]
#[serde(transparent)]
/// Converts `linearGradient` and `radialGradient` nodes that are a solid colour
/// to the equivalent colour.
///
/// # Errors
///
/// Never.
///
/// If this job produces an error or panic, please raise an [issue](https://github.com/noahbald/oxvg/issues)
pub struct ConvertOneStopGradients(pub bool);

#[derive_where(Default)]
struct State<'arena, E: Element<'arena>> {
    /// Parent defs with removed gradients
    effected_defs: RefCell<HashSet<E>>,
    /// All defs and their associated parent
    all_defs: RefCell<HashSet<E>>,
    gradients_to_detach: RefCell<HashSet<E>>,
    xlink_href_count: Cell<usize>,
    marker: PhantomData<&'arena ()>,
}

impl<'arena, E: Element<'arena>> Visitor<'arena, E> for ConvertOneStopGradients {
    type Error = String;

    fn prepare(
        &self,
        document: &E,
        info: &Info<'arena, E>,
        _context_flags: &mut ContextFlags,
    ) -> Result<PrepareOutcome, Self::Error> {
        if self.0 {
            State::default().start(&mut document.clone(), info, None)?;
        }
        Ok(PrepareOutcome::skip)
    }
}

impl<'arena, E: Element<'arena>> Visitor<'arena, E> for State<'arena, E> {
    type Error = String;

    fn prepare(
        &self,
        _document: &E,
        _info: &Info<'arena, E>,
        _context_flags: &mut ContextFlags,
    ) -> Result<PrepareOutcome, Self::Error> {
        Ok(PrepareOutcome::use_style)
    }

    fn element(
        &self,
        element: &mut E,
        context: &mut Context<'arena, '_, '_, E>,
    ) -> Result<(), Self::Error> {
        let xlink_attr = <E::Attr as Attr>::Name::new(Some("xlink".into()), "href".into());
        if element.has_attribute(&xlink_attr) {
            self.xlink_href_count.set(self.xlink_href_count.get() + 1);
        }
        if element.local_name().as_ref() == "defs" {
            self.all_defs.borrow_mut().insert(element.clone());
            return Ok(());
        }
        if element.local_name().as_ref() != "linearGradient"
            && element.local_name().as_ref() != "radialGradient"
        {
            return Ok(());
        }

        let mut stops = element
            .child_elements_iter()
            .filter(|child| child.prefix().is_none() && child.local_name().as_ref() == "stop");

        let href = element
            .get_attribute_node(&xlink_attr)
            .or_else(|| element.get_attribute_node_local(&"href".into()));
        let effective_node = if stops.next().is_none() {
            if let Some(href) = href {
                if href.value().starts_with('#') {
                    context
                        .root
                        .select(href.value())
                        .ok()
                        .and_then(|mut i| i.next())
                } else {
                    Some(element.clone())
                }
            } else {
                Some(element.clone())
            }
        } else {
            Some(element.clone())
        };

        let mut gradients_to_detach = self.gradients_to_detach.borrow_mut();
        let Some(effective_node) = effective_node else {
            log::debug!("no effective nodes for gradient");
            gradients_to_detach.insert(element.clone());
            return Ok(());
        };

        let effective_stops: Vec<_> = effective_node
            .child_elements_iter()
            .filter(|child| child.prefix().is_none() && child.local_name().as_ref() == "stop")
            .collect();

        if effective_stops.len() != 1 {
            log::debug!("skipping, multiple/zero stops for gradient");
            return Ok(());
        }

        if let Some(parent) = element.parent_element() {
            if parent.prefix().is_none() && parent.local_name().as_ref() == "defs" {
                self.effected_defs.borrow_mut().insert(parent);
            }
        }

        gradients_to_detach.insert(element.clone());

        let color = get_color(context, &effective_stops);
        let Some(id) = element.get_attribute_local(&"id".into()) else {
            log::debug!("skipping reference updates, no id");
            return Ok(());
        };
        log::debug!("updating colors: {color:?}");
        let selector_val = format!("url(#{id})");
        update_color_references(context, color.as_ref(), &selector_val)?;
        update_style_references(context, color.as_ref(), &selector_val)
    }

    fn exit_element(
        &self,
        element: &mut E,
        _context: &mut Context<'arena, '_, '_, E>,
    ) -> Result<(), Self::Error> {
        if element.prefix().is_some() || element.local_name().as_ref() != "svg" {
            return Ok(());
        }

        let xlink_href = <E::Attr as Attr>::Name::new(Some("xlink".into()), "href".into());
        for gradient in self.gradients_to_detach.borrow().iter() {
            if gradient.has_attribute(&xlink_href) {
                self.xlink_href_count.set(self.xlink_href_count.get() - 1);
            }

            gradient.remove();
        }

        if self.xlink_href_count.get() == 0 {
            element.remove_attribute(&<E::Attr as Attr>::Name::new(
                Some("xmlns".into()),
                "xlink".into(),
            ));
        }

        let effected_defs = self.effected_defs.borrow();
        for def in self.all_defs.borrow().iter() {
            if !def.has_child_elements() && effected_defs.contains(def) {
                def.remove();
            }
        }

        Ok(())
    }
}

fn get_color<'arena, E: Element<'arena>>(
    context: &mut Context<'arena, '_, '_, E>,
    effective_stops: &[E],
) -> Option<Result<E::Atom, String>> {
    let effective_stop = effective_stops.first().expect("len should be 1");
    let computed_styles = ComputedStyles::default().with_all(
        effective_stop,
        &context.stylesheet,
        context.element_styles,
    );

    let stop_color_unknown = PropertyId::Custom(CustomPropertyName::Unknown("stop-color".into()));
    // FIXME: Use `get_computed_styles_factory!`
    computed_styles
        .important_declarations
        .get(&stop_color_unknown)
        .map(|p| &p.1)
        .or_else(|| computed_styles.inline_important.get(&stop_color_unknown))
        .or_else(|| computed_styles.inline.get(&stop_color_unknown))
        .or_else(|| {
            computed_styles
                .declarations
                .get(&stop_color_unknown)
                .map(|p| &p.1)
        })
        .or_else(|| computed_styles.attr.get(&PresentationAttrId::StopColor))
        .or_else(|| computed_styles.inherited.get(&Id::CSS(stop_color_unknown)))
        .or_else(|| {
            computed_styles
                .inherited
                .get(&Id::Attr(PresentationAttrId::StopColor))
        })
        .and_then(|style| match style {
            Style::Static(style) => match style {
                Static::Css(Property::Custom(CustomProperty { value, .. })) => {
                    value.0.first().and_then(|token| match token {
                        TokenOrValue::Color(color) => Some(color),
                        _ => None,
                    })
                }
                Static::Attr(PresentationAttr::StopColor(color)) => Some(color),
                _ => None,
            },
            Style::Dynamic(_) => None,
        })
        .map(|color| {
            color
                .to_css_string(PrinterOptions::default())
                .map_err(|e| e.to_string())
                .map(Into::into)
        })
}

fn update_color_references<'arena, E: Element<'arena>>(
    context: &mut Context<'arena, '_, '_, E>,
    color: Option<&Result<E::Atom, String>>,
    selector_val: &str,
) -> Result<(), String> {
    let selector = COLORS_PROPS
        .iter()
        .map(|attr| format!(r#"[{attr}="{selector_val}"]"#))
        .join(",");
    let elements = match context.root.select(&selector) {
        Ok(elements) => elements,
        Err(err) => {
            log::debug!("unable to parse selector `{selector}`: {err:?}");
            return Ok(());
        }
    };
    for element in elements {
        for attr in &COLORS_PROPS {
            let attr_name = (*attr).into();
            let Some(mut attr) = element.get_attribute_node_local_mut(&attr_name) else {
                continue;
            };
            if attr.value().as_ref() != selector_val {
                continue;
            }

            if let Some(color) = color {
                attr.set_value(color.as_ref().map_err(ToString::to_string)?.clone());
            } else {
                drop(attr);
                element.remove_attribute_local(&attr_name);
            }
        }
    }
    Ok(())
}

fn update_style_references<'arena, E: Element<'arena>>(
    context: &Context<'arena, '_, '_, E>,
    color: Option<&Result<E::Atom, String>>,
    selector_val: &str,
) -> Result<(), String> {
    let styled_selector = format!(r#"[style*="{selector_val}"]"#);
    let styled_elements = match context.root.select(&styled_selector) {
        Ok(elements) => elements,
        Err(err) => {
            log::debug!("unable to parse selector `{styled_selector}`: {err:?}");
            return Ok(());
        }
    };
    let style = "style".into();
    for element in styled_elements {
        let Some(mut attr) = element.get_attribute_node_local_mut(&style) else {
            continue;
        };

        let color = match color {
            Some(Ok(ref color)) => color,
            Some(Err(err)) => return Err(err.to_string()),
            None => *AttrsGroups::Presentation
                .defaults()
                .unwrap()
                .get("stop-color")
                .unwrap(),
        };
        let color = attr.value().as_ref().replace(selector_val, color).into();
        attr.set_value(color);
    }
    Ok(())
}

#[test]
fn convert_one_stop_gradients() -> anyhow::Result<()> {
    use crate::test_config;

    insta::assert_snapshot!(test_config(
        r#"{ "convertOneStopGradients": true }"#,
        Some(
            r##"<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="744.09448" height="1052.3622">
  <!-- Convert both a one-stop gradient configured from attribute and styles. -->
  <defs>
    <linearGradient id="a">
      <stop stop-color="#ddc4cc"/>
    </linearGradient>
    <linearGradient id="b">
      <stop style="stop-color:#a8c4cc"/>
    </linearGradient>
  </defs>
  <rect width="150" height="150" x="350" y="350" fill="url(#a)"/>
  <rect width="150" height="150" x="50" y="350" style="fill:url(#b)"/>
</svg>"##
        ),
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "convertOneStopGradients": true }"#,
        Some(
            r##"<svg xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="744.09448" height="1052.3622">
  <!-- Convert a one-stop gradient that references another one-stop gradient. -->
  <!-- Remove xlink:href namespace since we remove the only reference to it. -->
  <defs>
    <linearGradient id="a">
      <stop style="stop-color:#a8c4cc"/>
    </linearGradient>
    <linearGradient x1="353.83112" y1="396.85037" x2="496.56262" y2="396.85037" id="b" xlink:href="#a"/>
  </defs>
  <rect width="150" height="150" x="350" y="350" style="fill:url(#b)"/>
</svg>"##
        ),
    )?);

    insta::assert_snapshot!(test_config(
        r#"{ "convertOneStopGradients": true }"#,
        Some(
            r##"<svg xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink" version="1.1" width="744.09448" height="1052.3622">
  <!-- If a one-stop gradient has the color defined via both attribute and style, style takes precedence. -->
  <defs>
    <linearGradient id="a">
      <stop stop-color="#ff0000" style="stop-color:#00ff00"/>
    </linearGradient>
    <linearGradient x1="353.83112" y1="396.85037" x2="496.56262" y2="396.85037" id="b" xlink:href="#a"/>
  </defs>
  <rect width="150" height="150" x="350" y="350" style="fill:url(#b)"/>
</svg>"##
        ),
    )?);

    Ok(())
}
