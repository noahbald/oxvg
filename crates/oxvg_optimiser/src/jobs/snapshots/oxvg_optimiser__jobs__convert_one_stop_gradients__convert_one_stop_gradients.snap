---
source: crates/oxvg_optimiser/src/jobs/convert_one_stop_gradients.rs
expression: "test_config(r#\"{ \"convertOneStopGradients\": true }\"#,\nSome(r##\"<svg xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\" width=\"744.09448\" height=\"1052.3622\">\n  <!-- Convert both a one-stop gradient configured from attribute and styles. -->\n  <defs>\n    <linearGradient id=\"a\">\n      <stop stop-color=\"#ddc4cc\"/>\n    </linearGradient>\n    <linearGradient id=\"b\">\n      <stop style=\"stop-color:#a8c4cc\"/>\n    </linearGradient>\n  </defs>\n  <rect width=\"150\" height=\"150\" x=\"350\" y=\"350\" fill=\"url(#a)\"/>\n  <rect width=\"150\" height=\"150\" x=\"50\" y=\"350\" style=\"fill:url(#b)\"/>\n</svg>\"##),)?"
---
<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="744.09448" height="1052.3622">
    <!-- Convert both a one-stop gradient configured from attribute and styles. -->
    <rect width="150" height="150" x="350" y="350" fill="#ddc4cc"/>
    <rect width="150" height="150" x="50" y="350" style="fill:#a8c4cc"/>
</svg>
